[3기-C 김환] 앨런팀 게시판 추가 미션 제출합니다

hwankim123 opened the pull request:
<!--
  템플릿은 아직 PR 작성이 익숙하지 않으신 분들을 위해서 제공하는 가이드입니다!
  리뷰어 또는 이 PR을 보게 될 다른 사람들이 이 PR을 보는데 참고할 수 있는 내용이 있다면 포함해서 작성해주시면 됩니다.
-->

## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->
- 게시판 과제에 회원기능을 추가하였습니다.
## 👩‍💻 요구 사항과 구현 내용 <!-- 기능을 Commit 별로 잘개 쪼개고, Commit 별로 설명해주세요 -->
- 회원가입기능
    - 사용자에게 이메일과 패스워드를 입력받아, 회원가입을 할 수 있습니다.
- 로그인 기능
    - 가입한 정보로 로그인을 할 수 있습니다.
    - 쿠키에 회원 id를 전달합니다.
- 내 정보 조회기능
    - 가입한 사용자는 자신의 정보를 조회할 수 있습니다.
    - 이 경우 로그인이 되어있어야하고, 본인만 정보를 조회할 수 있어야 합니다.
    - 쿠키의 회원 id를 통해 권한을 확인합니다.
## ✅ 피드백 반영사항  <!-- 지난 코드리뷰에서 고친 사항을 적어주세요. 재PR 시에만 사용해 주세요! (재PR 아닌 경우 삭제) -->
승훈님 추가 리뷰의 반영사항이 있습니다! 코멘트로 코드라인에 추가 설명 남기겠습니다!
- 연관관계 참조 객체를 반환하는 getter에 방어적 복사를 적용하였습니다.
- 불필요한 메서드를 없앴습니다.
- 테스트 코드의 코드 중복을 줄였습니다.
- Entity 수준에서 validation을 진행하였습니다. Bean Validation을 사용하였습니다.

## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->
회원 기능에 대한 response를 어떻게 내려주는 것이 좋을지 고민해봤는데, 가장 좋은 방법을 찾지 못하였습니다. 이 과정에서 겪은 고민들을 공유해드립니다!
- 로그인, 로그아웃 기능에 대한 적절한 status를 정하지 못하였습니다. 회원가입은 201로 하는 것이 적절하다는 결론을 내렸는데, 로그인과 로그아웃은 Post요청이긴 하나 리소스가 생성되지 않는 요청이기 때문에 고민이 생겼던 것 같습니다.
- 회원가입, 로그인, 로그아웃 응답 헤더에 location으로 url을 넘겨주는 것이 좋을까요? 아니면 location url을 넘겨주지 않고 프론트에게 응답 이후의 과정을 위임하는 것이 좋을까요?

hwankim123 committed:
feat: email과 password 필드 추가

hwankim123 committed:
feat: Member 엔티티의 연관관계 참조값인 posts List에 대한 getter에 방어적 복사 적용

hwankim123 committed:
refactor: 불필요한 메서드 제거

hwankim123 committed:
chore: spring-boot-starter-validation 의존성 추가

hwankim123 committed:
feat: Entity validation 기능 추가

hwankim123 committed:
refactor: 테스트 코드 리팩토링
- 중복 코드 제거
- 실제값에는 로직이 들어가지 않도록 수정
- 예상값에는 상수값이 들어가도록 수정

hwankim123 committed:
feat: 회원가입, 로그인, 마이페이지, 로그아웃 구현

- 이메일 비밀번호로 인증
- 쿠키로 권한 확인 후 마이페이지 조회

hwankim123 committed:
feat: UnAuthorizedException 추가

hwankim123 committed:
feat: 회원 도메인 테스트코드 작성

hwankim123 commented on src/main/java/com/example/board/domain/member/entity/Member.java:
```diff
@@ -13,23 +13,43 @@
   @Column(name = "name")
   private String name;
 
+  @NotNull
+  @Email
+  @Column(name = "email")
+  private String email;
+
+  @NotNull
+  @Length(min = 6, max = 20)
+  @Column(name = "password")
+  private String password;
+
+  @Min(value = 6)
+  @Max(value = 140)
   @Column(name = "age")
   private int age;
 
+  @Length(max = 20)
   @Column(name = "hobby")
   private String hobby;
```
Bean Validation으로 생성, 수정 시 값 검증을 진행합니다.

멘토님들이 Bean Validation을 Entity에 사용할 때의 문제점을 듣기 전에 작성한 코드입니다.
얘기를 듣고 나니 직접 validation 메서드를 작성하는 것도 고민되네요...

seung-hun-h replied:
저도 그때 검증에 대한 이야기를 나누면서 명확하지는 않았던 거 같아요.

지금은 엔티티 필드가 많지 않은데, 만약 더 많아지게 된다면 전체 필드에 대한 유효성 검증을 일일이 모두 한다면 번거로울 것 같다는 생각도 드네요.

Bean Validation 애너테이션을 사용한 검증은 조금 널널(?)하게 설정 해두고, 엔티티에 비즈니스 로직이 있을 경우에는 그 메서드에서 수동으로 조금 더 엄격하게 검증을 수행할 수 있을 것 같습니다.

hwankim123 replied:
충분히 생각해보고 기준을 갖고 있는게 중요할 것 같네요. 감사합니다!!

hwankim123 commented on src/main/java/com/example/board/domain/member/entity/Member.java:
```diff
@@ -67,6 +101,6 @@ public String getHobby(){
 
   public List<Post> getPosts(){
-    return posts;
+    return new ArrayList<>(posts);
```
posts 필드에 대한 getter에 방어적 복사를 적용하였습니다.

이렇게 해주니 `Post` 엔티티의 연관관계 편의 메서드 `registerMember()`에서 `posts`에 직접 접근해 `posts.add()` 메서드를 호출하던 코드가 자연스럽게 고쳐지게 되었습니다. 이제는 `Member.addPost()` 메서드에서 내부적으로 `posts.add()`를 호출합니다. 좋은 것 같습니다!

hwankim123 commented on src/main/java/com/example/board/domain/post/entity/Post.java:
```diff
@@ -57,7 +57,7 @@ public Post(String title, String content, Member member) {
     this.title = newTitle;
     this.content = newContent;
-    preUpdate(memberName);
+    this.updatedBy = memberName;
```
불필요한 메서드를 제거하였습니다.

hwankim123 commented on src/test/java/com/example/board/domain/member/entity/MemberTest.java:
```diff
@@ -6,78 +6,74 @@
 import org.junit.jupiter.api.Test;
 
 class MemberTest {
+  private static final String NAME = "김환";
+  private static final String EMAIL = "email123@naver.com";
+  private static final String PASSWORD = "password123!";
+  private static final int AGE = 25;
+  private static final String HOBBY = "게임";
```
중복되는 값들을 필드로 선언하였습니다.

hwankim123 commented on src/test/java/com/example/board/domain/post/entity/PostTest.java:
```diff
@@ -1,65 +1,82 @@
 import org.junit.jupiter.api.Test;
 
 class PostTest {
+  private static final String NAME = "김환";
+  private static final String EMAIL = "email123@naver.com";
+  private static final String PASSWORD = "password123!";
+  private static final int AGE = 25;
+  private static final String HOBBY = "게임";
+
+  private static final String TITLE = "RBF";
+  private static final String CONTENT = "RBF를 작성합니다";
+
+  private final Member member = new Member(NAME, EMAIL, PASSWORD, AGE, HOBBY);
```
중복되는 값과 member객체를 필드로 선언하고 재사용하도록 수정하였습니다.

seung-hun-h replied:
멤버에서만 사용하는 정보는 상수일 필요가 없지 않을까요?

hwankim123 replied:
Member class 초기화 시에 값들을 바로 넣어주면 된다는 말씀이시죠? 이 방식이 더 좋아보이네요!

hwankim123 reviewed (commented):
@seung-hun-h 
승훈님! 이전 추가 리뷰의 반영사항입니다!

haero77 commented on src/main/java/com/example/board/domain/member/controller/MemberController.java:
```diff
@@ -19,23 +25,19 @@ public class MemberController {
   private final MemberService memberService;
 
+  @Value("${set-cookie.max-age}")
+  private int maxAge;
```
yaml 을 이용해 쿠키 만료 시간을 세팅하셨군요! 
확실히 매직 넘버나 상수로 세팅하는 것보다 이 방식이 유지보수하기 좋을 것 같네요. 배워갑니다!👍

haero77 commented on src/main/java/com/example/board/domain/member/dto/MemberRequest.java:
```diff
@@ -1,5 +1,19 @@
+
+  }
+
+  public record SignUp(String name, String email, String password, int age, String hobby) {
+
+    public Member toEntity() {
+      return new Member(name, email, password, age, hobby);
+    }
+  }
```
현재 `Member` 인스턴스를 생성해주는 책임을 DTO가 갖고 있는 것으로 보입니다.
이렇게 되면 DTO가 추가될 때마다 `toEntity` 를 추가적으로 작성해주어야해서 조금 번거로울 수도 있겠네요.
인스턴스 생성에 대한 책임을 엔티티 객체가 갖도록 하면 DTO가 추가되어도 중복되는 코드를 줄일 수 있을 것 같은데 환님 생각은 어떠신가요?

hwankim123 replied:
DTO가 Entity 생성 책임까지 갖고 있는 것은, Entity(Domain 객체)는 최대한 다른 의존성을 갖지 않도록 설계하라는 의도를 따라간 것이었습니다.
Entity 요구사항(실질적인 도메인 요구사항)이 변경되어서 DTO도 같이 변경되는 것은 괜찮아 보였지만, DTO 설계의 미스로 Entity 코드까지 변화가 일어나야 한다면, 불필요한 관리 포인트가 늘어나는 느낌이었습니다.

선호님 생각은 어떠신가요?

haero77 replied:
현재 '회원을 생성한다' 라는 하나의 비즈니스 로직에 대한 책임을 완수하기 위해, 
DTO 객체가 여기에 참여하고 있을 뿐만 아니라, 엔티티 클래스의 인스턴스를 생성해주는 책임을 맡고 있어요. 
개인적으로는 DTO마다 이런 중요한 책임이 담겨있으면 안 된다고 생각해요.

`엔티티 객체 스스로 자신의 인스턴스 생성에 대한 책임을 갖도록 설계`함으로써, 
다른 의존성은 갖지 않고, 동시에 DTO 설계 미스로 인한 엔티티 코드의 변화를 방지할 수 있을 것 같습니다.

haero77 replied:
오늘(23.01.03) 스크럼을 통해

[`DTO.toEntity Vs. 서비스 레이어에서 DTO의 필드값 직접 사용`](https://www.notion.so/backend-devcourse/13-da89c205962e4e53bc3f70af8e1290eb#2797a6f6dc2e4f05b780cc5b32216c11) 에 대해 논의해보았고, 저 역시 많은 생각을 해보게 되었는데요.

- 객체지향세계에서는 모든 것이 객체이므로 DTO 역시 객체이다.
- `toEntity`를 사용하면 DTO 클래스만 변경하면 되므로 유지보수성이 후자에 비해 뛰어나다.
- 후자에 비해 더 직관적이다. 

등의 장점을 이유로 저 역시 `toEntity` 를 사용해 리팩토링 해봐야겠다고 생각이 변하였습니다.

또한, 수린(@Tnfls99)님께서 제시해주신 `toEntity에서 엔티티의 정적 팩터리 메서드를 호출하여 엔티티 인스턴스를 생성`하면 엔티티는 스스로의 인스턴스를 생성하는 책임을 유지할 수 있으므로 해당 방법으로 리팩토링 해봐야겠네요. 덕분에 많은 공부가 되었어요 감사합니다🥳 인사이트 나눠주신 수린님도 감사드려요🥳

haero77 commented on src/main/java/com/example/board/domain/member/dto/MemberResponse.java:
```diff
@@ -9,13 +9,15 @@ public class MemberResponse {
+  public record Detail(Long id, String name, String email, String password, int age, String hobby, List<PostResponse.Shortcut> posts,
+                       LocalDateTime createdAt, LocalDateTime updatedAt){
 
     public static Detail from(Member member) {
       return new Detail(
           member.getId(),
           member.getName(),
+          member.getEmail(),
+          member.getPassword(),
```
```suggestion
    public Detail(Member member) {
      this(member.getId(), member.getName(), ... );
    }
```

Detail 클래스에서 정적 팩터리 메서드를 이용해 인스턴스를 생성하고 계시군요!
별도의 메서드를 만들지 않고 `this()`를 활용해 대체해보는 것은 어떨까요? ☺️

hwankim123 replied:
정적 팩토리 메서드 대신 생성자를 사용해보는 것을 추천하신거죠? 혹시 그렇게 생각하신 이유를 여쭤봐도 될까요?

리뷰를 보고 알아보았는데, record는 모든 필드를 파라미터로 받는 생성자를 private으로 선언할 수 없더라구요.
이렇게 되면 생성자도 public으로 열리고, 정적 팩토리 메서드까지 있게 되어서, 이 DTO를 사용하는 입장에서는 무엇을 선택해야할지 혼란이 올수도 있다고 생각이 들었습니다.

그래서 정적 팩토리 메서드와 생성자 모두 public으로 열려있는 것 보단, 생성자 방식 하나로 통일하고 내부에서는 this()를 사용하는 것이 더 좋다고 생각이 들었습니다!
그래도 더 명확한 기준은 필요해보이는데, 같이 생각해볼까요? 선호님은 어떠신가요?

haero77 replied:
저는 OOP관점에서 'DTO 클래스에서 정적 팩토리 메서드를 사용하지 않아야한다.' 고 생각합니다.
그 이유는 'DTO 인스턴스를 만드는 책임과 인스턴스 생성 시점을 명확히하기 위함' 입니다.

먼저, 정적 팩터리 메서드를 이용하는 경우를 생각해볼게요.
- 서비스 객체는 DTO 객체에게 'DTO 인스턴스 생성해라' 라는 메시지를 전달하기 위해 DTO 클래스의 정적 팩터리 메서드를 호출하게 됩니다.
- DTO 객체는 자신에게 부여받은 'DTO 인스턴스 생성' 에 대한 책임을 위해 `from` 메서드에서 인스턴스를 생성하고, 해당 인스턴스의 주솟값만 서비스 레이어에게 리턴함으로써 책임을 완수합니다.
- 즉, DTO클래스에서 정적 팩터리 메서드를 사용하면, 모든 DTO 클래스는 자신의 인스턴스를 만들어야하는 책임이 생깁니다. 책임이 생겼으므로 모든 DTO클래스는 관리를 해줘야겠군요.


다음은 생성자를 이용하는 경우입니다.
- 서비스 객체는 'DTO 인스턴스를 생성하라'라는 책임을 스스로 갖습니다. 해당 책임을 완수하기 위해 단순히 DTO 클래스의 생성자를 호출하기만 하면 됩니다. 
- DTO 클래스에 부여된 책임은 전혀없으며, 따라서 설계된 의도(단순 데이터 전달용)대로 사용할 수 있게 됩니다. 생성자 오버로딩을 제외하면 DTO클래스에는 메서드가 전혀 없으니까요. 추가적으로 작성한 코드도 없으니, DTO 클래스를 관리하는 부담이 한결 줄었습니다.

상기 이유로, 저는 DTO 클래스에서 정적 팩터리 메서드를 사용하지 않습니다. 단순히 데이터 전송을 하기 위한 인스턴스 생성용 클래스인데, 별도의 책임을 갖는 것 자체가 부담이 되니까요. 물론 제가 아직 OOP에 관해 많이 부족하지만, 객체지향에 대해 공부하며 생긴 제 나름대로의 기준을 바탕으로 설명드려봤습니다☺️

hwankim123 replied:
오늘 스크럼 시간에 얘기를 나눠본 결과 "별도의 로직이 필요하지 않다면 생성자를 사용하자"쪽으로 결론을 내린 것 같습니다. 
상황에 따라 생성자 방식과 정적 팩토리 메서드 방식 중 무엇이 나을 지 고민해봐야겠네요. 
감사합니다!!

haero77 commented on src/main/java/com/example/board/domain/member/entity/Member.java:
```diff
@@ -13,23 +13,43 @@
   @Column(name = "name")
   private String name;
 
+  @NotNull
+  @Email
+  @Column(name = "email")
+  private String email;
```
정규식을 이용하여 이메일 포맷을 더욱 강력히 강제해보면 어떨까요?
[Email ID Validation in Spring boot](https://medium.com/@arunkumarmeenakshisundaram/email-id-validation-in-spring-boot-8c0fceebca79)을 참고해보시면 좋을 것 같습니다!

hwankim123 replied:
정규식 방식 한번 적용해보겠습니다! 
테스트 해보면서 어떤 부분까지 validate 가능한지 명확하지 않았는데, 찾아봐야겠군요. 감사합니다!

haero77 commented on src/main/java/com/example/board/domain/member/service/MemberService.java:
```diff
@@ -26,10 +27,27 @@ public MemberResponse.Detail findById(Long id) {
   }
+
+  public Long login(MemberRequest.Login loginRequest) {
+    Member member = memberRepository.findByEmail(loginRequest.email())
+        .orElseThrow(
+            () -> new UnAuthorizedException(
+                String.format("email %s not found", loginRequest.email())));
```
`findByEmail` 부분은 메서드로 분리해도 좋을 것 같습니다. 
저는 개인적으로 try-catch 문이나 예외를 발생시키는 로직은 별도의 메서드로 분리하려고 하는데, 
이렇게 함으로써 코드를 재사용하기 쉬워지더라고요!

hwankim123 replied:
좋은 것 같습니다!

다만 멘토님 리뷰에서 email과 password를 한번에 사용하여 DB 조회를 해보라는 리뷰가 있어서,
그 리뷰의 반영으로 인해 findByEmail이 필요가 없어질 수도 있겠네요...

한번 리뷰 반영해보면서 적용할 수 있을지 살펴보겠습니다.

haero77 commented on src/main/java/com/example/board/domain/member/dto/MemberRequest.java:
```diff
@@ -1,5 +1,19 @@
+
+  }
+
+  public record SignUp(String name, String email, String password, int age, String hobby) {
+
+    public Member toEntity() {
+      return new Member(name, email, password, age, hobby);
+    }
+  }
```
![image](https://user-images.githubusercontent.com/65555299/210091411-0f6decdd-501e-48c2-9355-eda937ad451e.png)

`age` 와 `hobby` 가 nullable 한 컬럼으로 확인되는데요, DTO `SignUp`은 age와 hobby 필드를 모두 받아야하네요.
사용자가 나이와 취미를 입력하지 않아도 회원가입을 할 수 있게 하신 것 같은데, 이를 충족시키도록 `SignUp` 을 수정해보신는 건 어떨까요?

hwankim123 replied:
RequestBody 어노테이션이 nullable 필드에 대해 다른 생성자를 알아서(?) 선택할 수 있다면 좋은 방법인 것 같네요!

그런데 애초에 RequestBody가 어떤 식으로 JSON을 역직렬화하는지에 대해 모르고 사용하는 것 같네요... 어노테이션을 사용하더라도 알고 사용해야 할 것 같습니다.

null인 상태로 돌아다니는 age, hobby 에 대한 처리를 확실히 해야할 것 같네요. 감사합니다!!

haero77 reviewed (commented):
@hwankim123 
환님 안녕하세요! 리뷰 맡은 김선호입니다!

`yaml` 을 활용한 쿠키 만료 시간 설정, ResponseCookie 를 활용하신 점이 인상 깊습니다! 저도 얼른 적용해보고 싶네요! 🥳

간단하게 리뷰를 남겼으니 확인해주시면 감사하겠습니다!

seung-hun-h commented on src/main/java/com/example/board/domain/member/controller/MemberController.java:
```diff
@@ -44,4 +46,33 @@ public ResponseEntity<Void> newMember(@RequestBody MemberRequest memberRequest)
+  public ResponseEntity<Void> login(@RequestBody MemberRequest.Login loginRequest) {
+    Long loginId = memberService.login(loginRequest);
+    ResponseCookie responseCookie = ResponseCookie.from("loginId", String.valueOf(loginId))
+        .httpOnly(Boolean.TRUE)
```
👍

seung-hun-h commented on src/main/java/com/example/board/domain/member/controller/MemberController.java:
```diff
@@ -44,4 +46,33 @@ public ResponseEntity<Void> newMember(@RequestBody MemberRequest memberRequest)
+  }
+
+  @GetMapping("/mypage")
+  public ResponseEntity<MemberResponse.Detail> mypage(@CookieValue(name = "loginId") Long memberId) {
```
우리 `HttpServletRequest`를 사용하기로 했던거 같네요

hwankim123 replied:
미션 요구사항을 놓쳤네요... HttpServletRequest를 이용하도록 수정해보겠습니다!

seung-hun-h commented on src/main/java/com/example/board/domain/member/controller/MemberController.java:
```diff
@@ -44,4 +46,33 @@ public ResponseEntity<Void> newMember(@RequestBody MemberRequest memberRequest)
+    return ResponseEntity.ok(detail);
+  }
+
+  @PostMapping("/logout")
+  public ResponseEntity<Void> logout(HttpServletResponse response) {
+    Cookie cookie = new Cookie("loginId", null);
+    cookie.setMaxAge(0);
+    response.addCookie(cookie);
+
+    return new ResponseEntity<>(HttpStatus.OK);
+  }
```
로그아웃을 쿠키를 삭제하는 것으로 구현 하셨네요. 좋은 것 같아요
`loginId` 쿠키 이름은 상수로 관리해주는 것이 좋겠네요. 이 메서드만 사용하는 것이 아니라 여러 메서드에서 사용하기 때문에 쿠키 이름이 변경되면 실수할 여지가 있을 것 같아요.

seung-hun-h commented on src/main/java/com/example/board/domain/member/service/MemberService.java:
```diff
@@ -26,10 +27,27 @@ public MemberResponse.Detail findById(Long id) {
+            () -> new UnAuthorizedException(
+                String.format("email %s not found", loginRequest.email())));
+
+    if(wrongPassword(loginRequest.password(), member.getPassword())){
+      throw new UnAuthorizedException(
+          String.format("password doesn't match. request: %s, actual: %s", loginRequest.password(), member.getPassword()));
+    }
```
`member.login(loginRequest.password());`
처럼 메서드를 호출할 수도 있겠네요

hwankim123 replied:
이부분은 고민했던 부분인데, 객체에 메세지를 전달함으로써 무엇에 더 집중하는 방식이 좋은 것 같네요. 반영해보겠습니다!

seung-hun-h commented on src/test/java/com/example/board/domain/member/controller/MemberControllerTest.java:
```diff
@@ -52,40 +53,69 @@ void tearDown(){
 
+  @Test
+  @DisplayName("이메일과 비밀번호로 로그인할 수 있습니다.")
+  void loginSuccess() throws Exception {
```
전체 테스트를 실행하면 이 테스트가 실패하네요.
원인을 파악해보고 해결해보면 좋을 것 같아요

hwankim123 replied:
Member와 연관된 Post 데이터들이 그대로 남아있어서 FK 제약조건에 걸려서 실패했군요...

`teardown()`에서 post 데이터도 모두 지우도록 하였습니다.

다만 PostRepository까지 의존성이 추가되니 뭔가 테스트가 무거워지는 것 같네요...

seung-hun-h reviewed (commented):
안녕하세요 환님 리뷰남겼습니다. 확인해주세요

---
- POST 메서드는 다른 메서드에 비해서 조금 더 범용적으로 사용합니다. 어떤 동작을 요청할 때도 POST 메서드를 사용하죠. 따라서 로그인이라는 행위가 자원을 생성하는 것은 아니지만 POST 메서드를 사용할 수 있겠죠. 개인적으로는 어떤 메서드를 사용할 지 애매할 때는 POST를 사용합니다. 
- 애매하지만 필요하다면 넘겨주는게 좋을 것 같네요.

hongbin-dev commented on src/main/java/com/example/board/domain/member/service/MemberService.java:
```diff
@@ -26,10 +27,27 @@ public MemberResponse.Detail findById(Long id) {
   }
+
+  public Long login(MemberRequest.Login loginRequest) {
+    Member member = memberRepository.findByEmail(loginRequest.email())
+        .orElseThrow(
```
지금상태에선 데이터베이스에서 필터링하면 어떨까요? `findByEmailAndPassword(...)`

hwankim123 replied:
처음엔 `password`까지 `WHERE`절에 넣음으로써 성능 최적화가 될수도 있겠구나 생각을 했는데, 다시 생각해보니 `email`컬럼에 Unique Constraint가 걸려있어서 의미있는 성능 최적화는 없을 것 같다고 생각이 되었습니다.

이 상태에서 `findByEmail`과 `findByEmailAndPassword`의 차이점은 한 개의 `member` row에 대해 `password` 비교까지 같이 수행하느냐의 차이인 것 같습니다.

제가 제대로 분석한게 맞을까요?


이와 별개로 DB 쿼리로 로그인 가능 여부를 바로 판단할 수 있어서 코드는 더 깔끔해지는 것 같네요!

hongbin-dev reviewed (commented):
환님 전반적으로 잘 작성해주셨군요 👍

hongbin-dev commented:
- 로그인, 로그아웃 기능에 대한 적절한 status를 정하지 못하였습니다. 회원가입은 201로 하는 것이 적절하다는 결론을 내렸는데, 로그인과 로그아웃은 Post요청이긴 하나 리소스가 생성되지 않는 요청이기 때문에 고민이 생겼던 것 같습니다.

REST API가 특히 로그인/로그아웃일때 더 모호한 생각이 들어요.
로그인 후 정보를 세션 정보를 생성하니 `POST`
세션정보를 삭제하니 `DELETE` 로 생각하면 자연스러울 거 같네요.

- 회원가입, 로그인, 로그아웃 응답 헤더에 location으로 url을 넘겨주는 것이 좋을까요? 아니면 location url을 넘겨주지 않고 프론트에게 응답 이후의 과정을 위임하는 것이 좋을까요?

이것도 팀과 협의가 우선순위겠지만,  넣어서 나쁠 건 없을 거 같네요. 
다만 해당 PATH가 변경된다면 location도 함께 변경되어야하는데요. 그 부분을 잘 챙겨야할 것 같네요

hwankim123 committed:
refector: Response DTO들의 정적 팩토리 메섣 from을 생성자로 변경

hwankim123 committed:
refector: PostRequest 리팩토링

- PostRequest로부터 Post를 생성하는 멤버 메서드 toEntity() 추가

hwankim123 committed:
refector: email 검증 정규식 추가

hwankim123 committed:
refactor: 권한 확인 로직 수정

- 미션 요구사항에 맞게 HttpServletRequest를 사용하도록 수정
- AuthorityUtils Util 클래스를 통해 Cookie 객체 추출

hwankim123 committed:
refactor: test 리팩토링

hwankim123 committed:
refactor: password 확인을 Member에서 수행하도록 수정

hwankim123 commented on src/main/java/com/example/board/common/util/AuthorityUtils.java:
```diff
@@ -0,0 +1,21 @@
+
+  private AuthorityUtils(){}
+
+  public static Cookie getCookie(HttpServletRequest request, String cookieName){
```
AuthorityUtils 유틸 클래스에서 Cookie 객체를 가져오는 getCookie 정적 메서드를 추가하였습니다.

hwankim123 commented on src/main/java/com/example/board/domain/member/controller/MemberController.java:
```diff
@@ -19,23 +27,22 @@ public class MemberController {
-    this.memberService = memberService;
-  }
+  @Value("${cookie.name}")
+  private String cookieName;
```
상수값 "loginId"를 application.yml 설정 파일에서 불러오도록 수정하였습니다.

seung-hun-h replied:
쿠키와 관련된 값들을 필드로 주입받고 있는데, 생성자에 주입 받도록 수정하면 좋을 것 같네요.

hwankim123 commented on src/main/java/com/example/board/domain/member/controller/MemberController.java:
```diff
@@ -44,4 +51,43 @@ public ResponseEntity<Void> newMember(@RequestBody MemberRequest memberRequest)
+  }
+
+  @GetMapping("/mypage")
+  public ResponseEntity<MemberResponse.Detail> mypage(HttpServletRequest request) {
+    Cookie loginIdCookie = AuthorityUtils.getCookie(request, cookieName);
+
+    try{
+      Long memberId = Long.valueOf(loginIdCookie.getValue());
+
+      MemberResponse.Detail detail = memberService.findById(memberId);
+
+      return ResponseEntity.ok(detail);
+    } catch(NumberFormatException e){
+      throw new UnAuthorizedException(
+          String.format("UnAuthorized cookie value: \"%s\"", loginIdCookie.getValue()));
+    }
+
+  }
```
HttpServletRequest로 cookie를 이용하도록 수정하였습니다.

hongbin-dev replied:
rethrowing 좋네요 👍

hwankim123 commented on src/main/java/com/example/board/domain/member/entity/Member.java:
```diff
@@ -38,12 +61,24 @@ public class Member extends BaseTimeEntity {
     this.hobby = hobby;
   }
 
+  public void login(String loginPassword) {
+    if(password.equals(loginPassword)){
+      throw new IllegalArgumentException("login failed. wrong password");
+    }
+  }
```
password 체크 로직을 Member의 메서드로 옮겼습니다.

seung-hun-h replied:
앨런 리뷰를 보고 저는 레포지토리에서 이메일과 패스워드를 조회하는 것이 더 좋다고 느꼈는데, 환님은 어떠신가요?

hwankim123 committed:
fix: password 확인 로직 수정

hwankim123 commented on src/test/java/com/example/board/domain/post/controller/PostControllerTest.java:
```diff
@@ -109,17 +109,17 @@ void findPostById() throws Exception {
     //given
-    MemberRequest memberRequest = new MemberRequest("김환", 25, "게임");
-    Long savedMemberId = memberService.save(memberRequest);
+    MemberRequest.SignUp signUpRequest = new MemberRequest.SignUp("김환", "email123@naver.com", "password123!", 25, "게임");
```
상수로 선언했던 값들을 모두 생성자에 바로 넘겨주도록 수정하였습니다.

hwankim123 reviewed (commented):
리뷰 반영해보았습니다! 추가 리뷰는 다음 미션에 커멘트 남기겠습니다!

seung-hun-h reviewed (commented):
환님 리뷰 남겼습니다~

hongbin-dev reviewed (approved):
환님 수고하셨습니다!
다음미션 진행하시죠
